test/utils.ts

```diff
@@ -97,6 +98,45 @@ export function test<T extends ValidTestCase>(
   }
 }
 
+type GetRuleType<TRule> =
+  TRule extends RuleModule<infer MessageIds, infer Options>
+    ? {
+        messageIds: MessageIds
+        options: Options
+      }
+    : never
+
+export function createRuleTestCaseFunction<
+  TRule extends RuleModule<string, unknown[]>,
+  TData extends GetRuleType<TRule> = GetRuleType<TRule>,
+  TTestCase extends
+    | TSESLintValidTestCase<TData['options']>
+    | TSESLintInvalidTestCase<TData['messageIds'], TData['options']> =
+    | TSESLintValidTestCase<TData['options']>
+    | TSESLintInvalidTestCase<TData['messageIds'], TData['options']>,
+>(): <
+  TReturn = TTestCase extends { errors: InvalidTestCaseError[] | number }
+    ? TSESLintInvalidTestCase<TData['messageIds'], TData['options']>
+    : TSESLintValidTestCase<TData['options']>,
+>(
+  t: TTestCase,
+) => TReturn {
```

Comment:
It is possible to re-use some code here for `export function test` as well? I.e. we could do `export const test = createRuleTestCaseFunction()` of sorts.

Reply:
Sure, I think that can be done, I'll give it a try!

Besides that my main idea is to replace the `export function test` with this new implementation:
If you are ok with `createRuleTestCaseFunction` approach I'll mark the old `test` function as deprecated.

Reply:
> It is possible to re-use some code here for export function test as well? I.e. we could do export const test = createRuleTestCaseFunction() of sorts.

Right now this is not feasible: if `test` function uses `createRuleTestCaseFunction`  the `errors` values will have to be of type `InvalidTestCase`.
This type requires `messageId` property (see Cons section in PR description), which result in 840 errors across 35 files 😅:

<img width="764" alt="image" src="https://github.com/user-attachments/assets/7c1143e1-f37c-4581-ae82-73d3c38d9057">

In the meantime I'll dedupe the code between the two functions.

Reply:
Oh I see, the `InvalidTestCase` from `@typescript-eslint/rule-tester` doesn't support the `message`. Maybe we could do this?

```ts
import type { InvalidTestCase as $InvalidTestCase } from `@typescript-eslint/rule-tester`

interface InvalidTestCase extends $InvalidTestCase {
  message?: string
}
```

Reply:
It should work, I don't see any problem theoretically. I'll give it a try!

Reply:
> IMHO after some tests is not worthy to loose time augmenting types just to deprecate the `message` property.
> If you agree I'll convert the errors with `messageId` alongside `createRuleTestCaseFunction` migration.

IMHO it is better to do that in a separate PR. There are a lot of `message` out there.

Reply:
> There are a lot of message out there.

Trust me, I've done worst 😆

> IMHO it is better to do that in a separate PR.

Ok, here is what I propose:
1. I'll finalise `createRuleTestCaseFunction` adding some JSDoc on this PR
2. deprecate old `test` function on this PR
3. If everything checks out you can merge this PR.
4. Like you told me is a lot of work so I'll consider to divide it into multiple PRs. 
    In each PRs I'll migrate one or more rules (if there are few tests),
    this way they can be easily reviewed and tested in CI and then merged.
    If I found other incorrect rule options types (like #174 and #175) I'll open a single PR for that rule.

Sounds like a plan?

Reply:
> Trust me, I've done worst 😆

That's not the worst 😆 Here is the worst: https://github.com/un-ts/eslint-plugin-import-x/pull/112

With the help of @merrywhether we have migrated almost all existing test cases.

> Sounds like a plan?

Sure, please go ahead!

Reply:
> That's not the worst 😆 Here is the worst: https://github.com/un-ts/eslint-plugin-import-x/pull/112
> With huge help from merrywhether, we had migrated all existing test cases with @typescript-eslint/rule-test@8.

There is always more space for the creating more "worst" scenarios when talking about programming 😆.
Nonetheless great work!

---

> Sure, please go ahead!

I pushed all changes related to point 1 and 2. If everything checks out you can merge this!